api/paidAction/index.js

```diff
@@ -45,248 +45,265 @@ export default async function performPaidAction (actionType, args, incomingConte
+  const optimistic = (me && supportedPaymentMethods.includes(PAID_ACTION_PAYMENT_METHODS.OPTIMISTIC))
 
-  const description = await paidActions[actionType].describe(args, contextWithSybilFeePercent)
-  const { invoice, wrappedInvoice, wallet, maxFee } = await createWrappedInvoice(userId, {
+  const { invoice, wrappedInvoice, wallet, maxFee, retriable } = await createWrappedInvoice(userId, {
     msats: cost,
     feePercent: sybilFeePercent,
     description,
-    expiry: INVOICE_EXPIRE_SECS
+    expiry: INVOICE_EXPIRE_SECS,
+    walletOffset: attempt
   }, { models, me, lnd })
 
-  const context = {
-    ...contextWithSybilFeePercent,
-    invoiceArgs: {
-      bolt11: invoice,
-      wrappedBolt11: wrappedInvoice,
-      wallet,
-      maxFee
-    }
+  context.invoiceArgs = {
+    bolt11: invoice,
+    wrappedBolt11: wrappedInvoice,
+    wallet,
+    maxFee
   }
 
-  return me
-    ? await performOptimisticAction(actionType, args, context)
-    : await beginPessimisticAction(actionType, args, context)
+  return {
+    retriable,
+    ...(optimistic
+      ? await performOptimisticAction(actionType, args, context)
+      : await beginPessimisticAction(actionType, args, context))
+  }
 }
 
-export async function retryPaidAction (actionType, args, incomingContext) {
-  const { models, me } = incomingContext
-  const { invoice: failedInvoice } = args
-
-  console.log('retryPaidAction', actionType, args)
-
-  const action = paidActions[actionType]
-  if (!action) {
-    throw new Error(`retryPaidAction - invalid action type ${actionType}`)
-  }
+export async function retryPaidAction ({ invoiceId, forceInternal, attempt, prioritizeInternal }, { ...context }) {
+  const { models, me } = context
 
   if (!me) {
-    throw new Error(`retryPaidAction - must be logged in ${actionType}`)
-  }
-
-  if (!action.paymentMethods.includes(PAID_ACTION_PAYMENT_METHODS.OPTIMISTIC)) {
-    throw new Error(`retryPaidAction - action does not support optimism ${actionType}`)
+    // otherwise every anon could retry retry other anons' actions
+    throw new Error('user must be logged in to retry paid actions')
   }
 
-  if (!action.retry) {
-    throw new Error(`retryPaidAction - action does not support retrying ${actionType}`)
-  }
+  const failedInvoice = await models.invoice.findUnique({ where: { id: invoiceId, userId: me.id } })
 
   if (!failedInvoice) {
-    throw new Error(`retryPaidAction - missing invoice ${actionType}`)
+    throw new Error('invoice not found')
   }
 
-  const { msatsRequested, actionId } = failedInvoice
-  const retryContext = {
-    ...incomingContext,
-    optimistic: true,
-    me: await models.user.findUnique({ where: { id: me.id } }),
-    cost: BigInt(msatsRequested),
-    actionId
+  if (failedInvoice.actionState !== 'FAILED') {
+    // you should cancel the invoice before retrying the action!
+    throw new Error(`actions is not in a retriable state: ${failedInvoice.actionState}`)
```

Comment:
Why does the client have to cancel the invoice? Can't we do it here?

Canceling on the client is an additional round-trip and the code might be simpler if retrying implicitly cancels the existing invoice.

What do you think?

Reply:
Hmm, I did consider implementing it like you've suggested, i've kept it this way because it is what the original code does.

But thinking about this in more depth, i think you are right, and that probably the transaction isolation should be serializable, because there is a race condition here (and in the original code), where two transactions could start with a committed 'failed' invoice state, but then end up with two different results.

What do you think?